### DISCUSSION TITLE
SPI w/ GPIO (non-CE)

### DISCUSSION THREAD

**Discussion by kev22257** [Pi4J/pi4j#264] [General]
Hi,

Cross post from [SO](https://stackoverflow.com/questions/75253056/pi4j-v2-spi-w-gpio-pin), but I'm trying to use v2 to talk to a BME280 with SPI and a GPIO pin.  I can't see how to configure the provider to do this, am I missing something?  I can't find an example and am not seeing what I need obviously in the javadoc.

Kevin

**Comment by FDelporte** [comment]
Is that a similar device as a BMP280?
https://github.com/Pi4J/pi4j-example-devices/blob/master/src/main/java/com/pi4j/devices/bmp280/README.md

**Reply by kev22257** [reply]
Yes, the BME additionally does humidity.  I've definitely looked at that page before, but it's I2C and not SPI.  Once I can get the device talking I'm sure that will be a valuable resource, thanks!

**Comment by taartspi** [comment]
Some notes first
https://github.com/Pi4J/pi4j-example-devices  Has three chips using SPI
ads1256/ADS1256.java:                
dac8552/DAC8552.java:                
mcp3008/MCP3008.java:      

If your config question pertains to the MODE. see https://abyz.me.uk/rpi/pigpio/python.html, search for mode.  Some datasheets require careful reading to determine the required MODE.

You mentioned using a GPIO for chip enable.
See dac8552/DAC8552.java.  This chip required its CS be held over several SPI transactions so the PIGPIO implementation CS was not usable.  Or at least I could not see how it could work.
Let us know if this answers your question.

**Reply by kev22257** [reply]
Yes, I think this is what I was looking for.  So SPI to talk to the device and a DO for the CE pin to make it talk.  I will put these together and see what I can do.  I really appreciate everyone taking the time to reply!

**Comment by kev22257** [comment]
Well looking at the DAC8552 example was the key.  Here is a working version that asks the device for it's ID (register 0xD0) and gives the correct answer for BME280 (0x60).  Thanks for the assist.  Here is the working code (needs to be cleaned up, but it's a basic start).

```
package com.pi4j.example.spi;

import com.pi4j.Pi4J;
import com.pi4j.context.Context;
import com.pi4j.io.gpio.digital.DigitalOutput;
import com.pi4j.io.gpio.digital.DigitalState;
import com.pi4j.io.spi.Spi;
import com.pi4j.io.spi.SpiBus;
import com.pi4j.io.spi.SpiChipSelect;
import com.pi4j.io.spi.SpiMode;
import com.pi4j.util.Console;
import com.pi4j.util.StringUtil;

public class BmeTest {

	private static final SpiBus spiBus = SpiBus.BUS_0;

	public static void main(String[] args) throws Exception {
		final var console = new Console();

		// print program title/header
		console.title("<-- The Pi4J Project -->",
				"Basic SPI Communications Example");

		Context pi4j = Pi4J.newAutoContext();
		var spiConfig = Spi.newConfigBuilder(pi4j)
				.id("SPI" + spiBus + "_BME280")
				.name("BME280")
				.bus(spiBus)
				.chipSelect(SpiChipSelect.CS_0) // not used
				.mode(SpiMode.MODE_0)
				.provider("pigpio-spi")
				.build();

		// required all configs
		var outputConfig = DigitalOutput.newConfigBuilder(pi4j)
				.id("CS_GPIO5")
				.name("CS GPIO5")
				.address(5)
				.shutdown(DigitalState.HIGH)
				.initial(DigitalState.HIGH)
				.provider("pigpio-digital-output");

		DigitalOutput csGpio = null;
		try {
			csGpio = pi4j.create(outputConfig);
		} catch (Exception e) {
			e.printStackTrace();
			console.println("create DigOut DRDY failed");
			System.exit(202);
		}

		// use try-with-resources to auto-close SPI when complete
		try (var spi = pi4j.create(spiConfig)) {
			byte data[] = new byte[] { (byte) (0x80 | 0xD0), (byte) 0x00 };

			csGpio.high();
			csGpio.low();
			spi.transfer(data, data);
			csGpio.high();

			// take a breath to allow time for the SPI
			// data to get updated in the SPI device
			Thread.sleep(100);

			// read data back from the SPI channel
			console.println("--------------------------------------");
			console.println("SPI [READ] :");
			console.println("  [BYTES]  0x" + StringUtil.toHexString(data));
			console.println("  [STRING] " + new String(data));
			console.println("--------------------------------------");
		}

		// shutdown Pi4J
		console.println("ATTEMPTING TO SHUTDOWN/TERMINATE THIS PROGRAM");
		pi4j.shutdown();
	}
}

```

> [main] INFO com.pi4j.util.Console - ************************************************************
> [main] INFO com.pi4j.util.Console - ************************************************************
> [main] INFO com.pi4j.util.Console -
> [main] INFO com.pi4j.util.Console -                   <-- The Pi4J Project -->
> [main] INFO com.pi4j.util.Console -               Basic SPI Communications Example
> [main] INFO com.pi4j.util.Console -
> [main] INFO com.pi4j.util.Console - ************************************************************
> [main] INFO com.pi4j.util.Console - ************************************************************
> [main] INFO com.pi4j.util.Console -
> [main] INFO com.pi4j.Pi4J - New auto context
> [main] INFO com.pi4j.Pi4J - New context builder
> [main] INFO com.pi4j.platform.impl.DefaultRuntimePlatforms - adding platform to managed platform map [id=raspberrypi; name=RaspberryPi Platform; priority=5; class=com.pi4j.plugin.raspberrypi.platform.RaspberryPiPlatform]
> [main] INFO com.pi4j.util.Console - --------------------------------------
> [main] INFO com.pi4j.util.Console - SPI [READ] :
> [main] INFO com.pi4j.util.Console -   [BYTES]  0xFF 60
> [main] INFO com.pi4j.util.Console -   [STRING] �`
> [main] INFO com.pi4j.util.Console - --------------------------------------
> [main] INFO com.pi4j.util.Console - ATTEMPTING TO SHUTDOWN/TERMINATE THIS PROGRAM

**Reply by FDelporte** [reply]
Thanks a lot for sharing your solution!!

**Reply by FDelporte** [reply]
BTW, can you add this as a solution to your SO question? Maybe link to this discussion? Thanks!

**Reply by kev22257** [reply]
Done, thanks again!
